virtool/samples/api.py

```diff
@@ -933,27 +926,25 @@ async def upload_artifacts_cache(req):
 
     name = req.query.get("name")
 
+    cache_path = virtool.caches.utils.join_cache_path(req.app["settings"], key) / name
+
     if artifact_type and artifact_type not in ArtifactType.to_list():
         return bad_request("Unsupported sample artifact type")
 
-    artifact = await virtool.samples.files.create_artifact_file(pg, name, sample_id, artifact_type, cache=True)
-
-    upload_id = artifact["id"]
-
-    cache_path = cache_path / artifact["name_on_disk"]
-
     try:
         size = await virtool.uploads.utils.naive_writer(req, cache_path)
     except asyncio.CancelledError:
-        logger.debug(f"Artifact file upload aborted: {upload_id}")
+        logger.debug(f"Artifact file cache upload aborted: {sample_id}")
```

Comment:
I decided to do the actual uploading before the row is created to avoid the need to use `finalize()` and having to remove the row if `CancelledError` is thrown. I lose access to `upload_id` as a result.

Reply:
What happens if two files with the same `name` are uploaded concurrently? Whichever one finishes first will get remembered in the database and the other will fail because of our naming constraint.

Thoughts?

Reply:
That makes sense, it will probably be safer to create the row first as the second file will return a conflict response before it has a chance to upload. Do you think a filesystem check before uploading would work just as well? If not I think I'll just revert to using `finalize()`.

Reply:
Stick to the database.